### PR TITLE
sql: improve crdb_internal.force_retry()

### PR DESCRIPTION
**Pull request by andreimatei**
Make it bump the epoch of the txn so that, after the force_retry() is
called, the SQL txn doesn't see its previous writes.

Release note: None

### CODE REVIEW

**Comment by nvanbenschoten** [comment]
@andreimatei does this sufficiently improve `crdb_internal.force_retry()` enough to close https://github.com/cockroachdb/cockroach/issues/16827?

**Comment by andreimatei** [comment]
indeed it does. Closing that issue.

File: pkg/sql/sem/builtins/builtins.go [hunk 1]
```diff
@@ -2276,6 +2276,8 @@ CockroachDB supports the following flags:
 					Nanos: int64(ctx.StmtTimestamp.Sub(ctx.TxnTimestamp)),
 				}
 				if elapsed.Compare(minDuration) < 0 {
+					ctx.Txn.Proto().Restart(ctx.Txn.UserPriority(), 0, /* upgradePriority */
+						ctx.Txn.Proto().Timestamp)
```
Comment (knz) [inline pkg/sql/sem/builtins/builtins.go]: I guess this code is redundant with the one in `GenerateForcedRetryableError`. You can remove it?

**Comment by andreimatei** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/sem/builtins/builtins.go, line 2280 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21836#-L3oMP94rCRAPNgqe9y6-r1-2280:-L48XU5hwo_VMxaRAnFJ:b8lyh5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d67e38e8e945aeb20172c45bc0e988e981680a2/pkg/sql/sem/builtins/builtins.go#L2280)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I guess this code is redundant with the one in `GenerateForcedRetryableError`. You can remove it?
</blockquote></details>

done, thanks. I had previously hesitated about where to put it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21836)*

**Review by knz — APPROVED** [review]

**Review by knz — APPROVED** [review]

**Pull request merged**